build.sbt

```diff
@@ -49,10 +49,48 @@ val commonSettings: immutable.Seq[Def.Setting[_]] = List(
     specsScalaCheck,
     specsMock
   ),
-  assembly / assemblyMergeStrategy := {
+  ThisBuild / assemblyMergeStrategy := {
     case "META-INF/MANIFEST.MF" => MergeStrategy.discard
-    case "META-INF/org/apache/logging/log4j/core/config/plugins/Log4j2Plugins.dat" =>
-      new MergeFilesStrategy
+    case PathList(ps @ _*) if ps.last equalsIgnoreCase "Log4j2Plugins.dat" =>
+      import java.io.FileInputStream
+      import java.io.FileOutputStream
+      import org.apache.logging.log4j.core.config.plugins.processor.PluginCache
+
+      import scala.collection.JavaConverters.asJavaEnumerationConverter
+
+      import sbt.io.{IO, Using}
+      import sbtassembly.Assembly.Dependency
+
+      CustomMergeStrategy("Log4j2Plugins") { conflicts =>
+        val dependencyStreamResource =
+          Using.resource((dependency: Dependency) => dependency.stream())
+        val tempDir = "target/log4j2-plugin-cache"
+        val urls = conflicts
+          .map { conflict =>
+            dependencyStreamResource(conflict) { is =>
+              val file = new File(
+                s"$tempDir/${conflict.module.get.jarName}-Log4j2Plugins.dat"
+              )
+              IO.write(file, IO.readBytes(is))
+              file.toURI.toURL
+            }
+          }
+        val aggregator = new PluginCache()
+        aggregator.loadCacheFiles(urls.toIterator.asJavaEnumeration)
+        val pluginCache = new File(s"$tempDir/Log4j2Plugins.dat")
+        val pluginCacheOutputStream = new FileOutputStream(pluginCache)
+        aggregator.writeCache(pluginCacheOutputStream)
+        pluginCacheOutputStream.close()
+        Right(
+          Vector(
+            JarEntry(
+              conflicts.head.target,
+              () => new FileInputStream(pluginCache)
+            )
+          )
+        )
+      }
+
```

Comment:
I guess this chunk of code has come from `project/MergeFilesStrategy.scala`, and taken on some more modifications?

 It's quite a big and obscure bit of code though, so I feel like keeping it in its own file in `MergeFilesStrategy.scala` is not a bad idea - we should probably include some explanation or references for it (like, where did we get this code, and what prompted the changes to it that are occurring now?)